### PR TITLE
DISTX-546 Implement cloud provider load-balancer support for datahub …

### PR DESCRIPTION
**Pull request by cegganesh84**
…cluster

1. This allows explicitly setting up a load-balancer to any datahub cluster
2. This reuses the load-balancer implementation for semi-private networks
3. When the load-balancer is enabled removed the restriction of one gateway node because this will be used by DE HA cluster and clusters with custom HA definition
4. If the load-balancer is enabled in the environment, then it is not possible to disable the load-balancer for a particular datahub cluster.
5. Modified 7.2.8 only for now to do the testing and getting the review in parallel, while I hash out the backward incompatibility in the endpoint for the released versions.

Tested with 7.2.7 HA cluster definition in the private stack that the load balancer is created. Also tested if it is not in the cluster definition the load balancer is not created

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java [hunk 1]
```diff
@@ -99,9 +100,31 @@ public StackV4Request convert(DistroXV1Request source) {
...
 
+    private void checkMultipleGatewayNodes(DistroXV1Request distroXV1Request) {
+        if (distroXV1Request.getInstanceGroups() == null || distroXV1Request.getInstanceGroups().isEmpty()) {
+            return;
+        }
+        distroXV1Request.getInstanceGroups().forEach(ig -> {
+            if (InstanceGroupType.GATEWAY == ig.getType()) {
+                if (distroXV1Request.isEnableLoadBalancer()) {
+                    if (ig.getNodeCount() < 1) {
+                        throw new BadRequestException("Instance group with GATEWAY type must contain at least 1 node!");
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java:116]: how is loadbalancing possible with 1 node?

File: core/src/main/resources/defaults/clustertemplates/7.2.8/aws/dataengineering-ha.json [hunk 2]
```diff
@@ -11,10 +11,11 @@
     "externalDatabase": {
       "availabilityType": "HA"
     },
+    "enableLoadBalancer": true,
     "instanceGroups": [{
       "nodeCount": 1,
       "name": "manager",
-      "type": "GATEWAY",
+      "type": "CORE",
```
Comment (bergerdenes) [inline core/src/main/resources/defaults/clustertemplates/7.2.8/aws/dataengineering-ha.json:18]: are you sure to run with 1 CORE node?

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java (hunk 1 above)
Comment (cegganesh84) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java:116]: @bergerdenes Cloud load-balancing solution is used in multiple contexts in Cloudbreak. It acts as a static endpoint. If you move CM to a different node and create the cluster with the same name, you still get the same endpoint. In that context allowing a single node GATEWAY with load-balancer is a useful feature. 

The other contexts are,

1. Semi-private networks where the load-balancer is hosted in public network/IP, and the VMs are hosted in private network/IP.
2. Loadbalancing with more than one node, to achieve resiliency in case of node failures. Like Hive endpoint.

**Comment by cegganesh84** [comment]
@bergerdenes @hreeve-cloudera @doktoric Could you please kindly take a look at this.

**Comment by cegganesh84** [comment]
A corresponding PR for thunderhead to enable the load-balancer via CLI https://github.infra.cloudera.com/thunderhead/thunderhead/pull/3677

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java [hunk 3]
```diff
@@ -146,12 +146,16 @@ public String getLoadBalancerUserFacingFQDN(Long stackId) {
         return loadBalancerOptional;
     }
 
-    public Set<LoadBalancer> createLoadBalancers(Stack stack, DetailedEnvironmentResponse environment) {
+    public Set<LoadBalancer> createLoadBalancers(Stack stack, DetailedEnvironmentResponse environment, boolean loadBalancerDefined) {
```
Comment (hreeve-cloudera) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:149]: Could we rename `loadBalancerDefined`? I don't think it's very descriptive, because at this point a load balancer isn't defined. Maybe something like `loadBalancerFlagEnabled`?

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java [hunk 4]
```diff
@@ -7,6 +7,7 @@
 
 import javax.inject.Inject;
 
+import com.sequenceiq.common.api.type.InstanceGroupType;
```
Comment (hreeve-cloudera) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java:10]: nit: This import is in the wrong place. It should be moved down with the rest of the `import com.*` below.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java [hunk 5]
```diff
@@ -99,9 +100,31 @@ public StackV4Request convert(DistroXV1Request source) {
...
+        checkMultipleGatewayNodes(source);
         return request;
     }
 
+    private void checkMultipleGatewayNodes(DistroXV1Request distroXV1Request) {
+        if (distroXV1Request.getInstanceGroups() == null || distroXV1Request.getInstanceGroups().isEmpty()) {
+            return;
+        }
+        distroXV1Request.getInstanceGroups().forEach(ig -> {
+            if (InstanceGroupType.GATEWAY == ig.getType()) {
```
Comment (hreeve-cloudera) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java:113]: The current load balancer logic already checks for the existence of the Knox service, and will only create load balancers if Knox is defined either explicitly or implicitly. Is there a reason not related to Knox we're checking the instance type here?

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java [hunk 6]
```diff
@@ -45,9 +44,6 @@
 
     private InstanceGroupV4Request convert(InstanceGroupV1Request source, DetailedEnvironmentResponse environment) {
         InstanceGroupV4Request response = new InstanceGroupV4Request();
-        if (InstanceGroupType.GATEWAY == source.getType() && source.getNodeCount() != 1) {
```
Comment (hreeve-cloudera) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java:48]: Why was this changed?

File: core/src/test/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverterTest.java [hunk 7]
```diff
@@ -17,6 +18,10 @@
 import java.util.Map;
 import java.util.Set;
 
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.base.RecoveryMode;
```
Comment (hreeve-cloudera) [inline core/src/test/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverterTest.java:21]: nit: Reorder imports.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java (hunk 5 above)
Comment (cegganesh84) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/DistroXV1RequestToStackV4RequestConverter.java:113]: Yes, the datahub has a special check not to allow GATEWAY node type to be defined for more than one node. This change basically enables to relax that check when a load-balancer is defined. This will enable us to have a master group with two nodes as GATEWAY where in the future we will also have CM HA.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java (hunk 6 above)
Comment (cegganesh84) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/converter/InstanceGroupV1ToInstanceGroupV4Converter.java:48]: This check was moved to a place where we can check if it is a load-balancer enabled cluster or not.

**Review by bergerdenes — APPROVED** [review]

**Comment by cegganesh84** [comment]
Thanks for the reviews folks. Could this be merged?

**Comment by bergerdenes** [comment]
Please have at least another pair of eyes check this PR

**Review by hreeve-cloudera — APPROVED** [review]

**Comment by cegganesh84** [comment]
There was a merge conflict from another load-balancer-related change. I have rebased and the checks are passing again.
